registry/storage/catalog.go

```diff
@@ -3,45 +3,50 @@ package storage
 import (
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"path"
+	"sort"
 	"strings"
 
 	"github.com/distribution/distribution/v3/reference"
 	"github.com/distribution/distribution/v3/registry/storage/driver"
 )
 
-// Returns a list, or partial list, of repositories in the registry.
+var (
+	// ErrStopReposWalk is used as a return value to indicate that the repository path walk
+	// should be stopped. It's not returned as an error by any function.
+	ErrStopReposWalk = errors.New("stop repos walk")
+)
+
+// Returns a list or a partial list of repositories in the registry.
 // Because it's a quite expensive operation, it should only be used when building up
 // an initial set of repositories.
 func (reg *registry) Repositories(ctx context.Context, repos []string, last string) (n int, err error) {
 	var finishedWalk bool
 	var foundRepos []string
 
 	if len(repos) == 0 {
```

Comment:
```suggestion
	if len(repos) == 0 && cap(repos) == 0 {
```

Otherwise, we'll error out on zero length slices that have a capacity such as one initialized like `var s = make([]string, 0, 100)`